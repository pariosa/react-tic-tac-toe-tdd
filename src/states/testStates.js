export const testStateNonEmpty = {
    ticTacGrid:[
        [
            {row:"0", col:"0",value:"X"},
            {row:"0", col:"1",value:""},
            {row:"0", col:"2",value:"O"}
        ],
        [
            {row:"1", col:"0",value:""},
            {row:"1", col:"1",value:"X"},
            {row:"1", col:"2",value:""}
        ],
        [
            {row:"2", col:"0",value:"O"},
            {row:"2", col:"1",value:""},
            {row:"2", col:"2",value:"X"}
        ],
    ],
    result:undefined,
    options:{
        playerSymbol:"X"
    },
    gameOver:false
};