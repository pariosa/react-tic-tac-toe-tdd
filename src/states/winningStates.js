export default function winningState(state){
    let winFlag = 0;
    let winTeam = "";
     
    //if one row has all the same val
    state.ticTacGrid.forEach((row)=>{
        if((row[0].value === row[1].value)&& (row[0].value === row[2].value) && (row[0].value !== "")){
            winFlag = 1;
            winTeam = row[0].value
        }
    }); 

    //if one col has all the same val
    let result = [];
    for(let i = 0; i < 3 ; i++){
        state.ticTacGrid.forEach(e => {
          result.push(e[i]);
        });
    };

    if((result[0].value === result[1].value) && (result[0].value === result[2].value) && (result[0].value !== "")){
        winFlag = 1;
        winTeam = result[0].value
    }
    if((result[3].value === result[4].value) && (result[3].value === result[5].value) && (result[3].value !== "")){
        winFlag = 1;
        winTeam = result[3].value
    }
    if((result[6].value === result[7].value) && (result[6].value === result[8].value) && (result[6].value !== "")){
        winFlag = 1;
        winTeam = result[6].value
    }  
  

    //tie status, all slots are non ""
    let emptyFlag = 0
    result.forEach(e=>{ if(e.value !== "O" || e.value !=="X"){emptyFlag += 1}})
    

    //if(row(0) col(0), row(1) col(1), and row(2) col(2) are all the same val
        const conA = state.ticTacGrid[0][0].value;
        const conB = state.ticTacGrid[1][1].value;
        const conC = state.ticTacGrid[2][2].value; 
        if((conA === conB) && (conA === conC) && (conA !== "")){
            winFlag = 1;
            winTeam = conA;
        }   


    //if row(0) col(2), row(1) col(1), and row(2), col(0) are all the same val
        const con1 = state.ticTacGrid[0][2].value;
        const con2 = state.ticTacGrid[1][1].value;
        const con3 = state.ticTacGrid[2][0].value;
        if((con1 === con2) && (con1 === con3) && (con1 !== "")){
            winFlag = 1;
            winTeam = con1;
        } 
        if (emptyFlag === 0){
            return(true,"Tie");
        }
         
        //return winFlag === 1 ? true : false;
        return (winFlag, winTeam);
};