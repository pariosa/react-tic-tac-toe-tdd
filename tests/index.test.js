import React from 'react';
import { render } from '@testing-library/react';
import Layout from '../src/index';
import '@testing-library/jest-dom';
import { shallow, mount} from "enzyme";
import initialState from "../src/states/initialState"; 
import { testStateNonEmpty } from "../src/states/testStates";
jest.mock("../src/TicTacToeSquare/TicTacToeSquare.css", () => {
    return {};
  });
jest.mock("../src/TicTacToeGrid/TicTacToeGrid.css", () => {
    return {};
});
jest.mock("../src/index.css", () => {
    return {};
});
const indexSnap = shallow(<Layout state={initialState} />);

it('renders shallow match', ()=>{
    expect(indexSnap).toMatchSnapshot();
})

test('renders index element', ()=>{
    const { getByText } = render(<Layout state={initialState}/>);
    const title = getByText(/Tic Tac Toe/i);
    expect(title).toBeInTheDocument();
});

it("should change the state of the component when an empty space is clicked", () =>{
    const wrapper = mount(<Layout {...initialState} />); 
    wrapper.find('.tic-tac-square').at(0).simulate('click');
    expect(wrapper.find('.tic-tac-square').at(0).prop("value")).toBe("X");
});

it('check to see if output of grid matches state object in state with data', () => {  
    const testLayoutGrid = mount(<Layout {...testStateNonEmpty} />);
    testLayoutGrid.find('.tic-tac-square').at(8).simulate('click');
    testLayoutGrid.update();
    expect(testLayoutGrid.find(".tic-tac-square").at(8).prop("value")).toBe(testStateNonEmpty.ticTacGrid[2][2].value);
});

it('check to see if output of grid matches state object in state with data', () => {  
    const testLayoutGrid = mount(<Layout {...testStateNonEmpty} />); 
    testLayoutGrid.find('.tic-tac-square').at(8).simulate('click');
    testLayoutGrid.update();
    expect(testLayoutGrid.find(".tic-tac-square").at(8).prop("value")).toBe(testStateNonEmpty.ticTacGrid[2][2].value);
});

it('checks a state where player O wins', ()=>{
    const testOGrid = mount(<Layout {...initialState} />);
    testOGrid.find('.tic-tac-square').at(2).simulate('click');
    testOGrid.find('.tic-tac-square').at(0).simulate('click');
    testOGrid.find('.tic-tac-square').at(1).simulate('click');
    testOGrid.find('.tic-tac-square').at(3).simulate('click');
    testOGrid.find('.tic-tac-square').at(4).simulate('click'); 
    testOGrid.find('.tic-tac-square').at(6).simulate('click'); 
     expect(testOGrid.find(".result").find('h2').prop("result")).toBe("O");
});

it('checks a state where player X wins', ()=>{
    const testXGrid = mount(<Layout {...initialState} />);
    testXGrid.find('.tic-tac-square').at(0).simulate('click');
    testXGrid.find('.tic-tac-square').at(1).simulate('click');
    testXGrid.find('.tic-tac-square').at(4).simulate('click');
    testXGrid.find('.tic-tac-square').at(5).simulate('click');
    testXGrid.find('.tic-tac-square').at(8).simulate('click'); 
    //console.log(testXGrid.debug());  
    expect(testXGrid.find(".result").find('h2').prop("result")).toBe("X");
});

it('checks a draw state',() => {
    const testDrawGrid = mount(<Layout {...initialState} />);
    //expect(testLayoutGrid.hasClass('tic-tac-grid')).toEqual(true);
    testDrawGrid.find('.tic-tac-square').at(0).simulate('click');
    testDrawGrid.find('.tic-tac-square').at(1).simulate('click');
    testDrawGrid.find('.tic-tac-square').at(2).simulate('click');
    testDrawGrid.find('.tic-tac-square').at(6).simulate('click');
    testDrawGrid.find('.tic-tac-square').at(7).simulate('click');
    testDrawGrid.find('.tic-tac-square').at(8).simulate('click');
    testDrawGrid.find('.tic-tac-square').at(3).simulate('click');
    testDrawGrid.find('.tic-tac-square').at(4).simulate('click');
    testDrawGrid.find('.tic-tac-square').at(5).simulate('click');
    //console.log(testDrawGrid.debug());  
    expect(testDrawGrid.find(".result").find('h2').prop("result")).toBe("TIE");
 });

it('successfully tie the game, rewind time the same amount of moves, and expect the intial state befoere moves were made',() => {
    const testRewindGrid = mount(<Layout {...initialState} />);  
    testRewindGrid.find('.tic-tac-square').at(0).simulate('click');
    testRewindGrid.find('.tic-tac-square').at(1).simulate('click');
    testRewindGrid.find('.tic-tac-square').at(2).simulate('click');
    testRewindGrid.find('.tic-tac-square').at(6).simulate('click');
    testRewindGrid.find('.tic-tac-square').at(7).simulate('click');
    testRewindGrid.find('.tic-tac-square').at(8).simulate('click');
    testRewindGrid.find('.tic-tac-square').at(3).simulate('click');
    testRewindGrid.find('.tic-tac-square').at(4).simulate('click');
    testRewindGrid.find('.tic-tac-square').at(5).simulate('click');
    expect(testRewindGrid.find(".result").find('h2').prop("result")).toBe("TIE");
    testRewindGrid.find('.rewind').at(0).simulate('click');
    testRewindGrid.find('.rewind').at(0).simulate('click');
    testRewindGrid.find('.rewind').at(0).simulate('click');
    testRewindGrid.find('.rewind').at(0).simulate('click');
    testRewindGrid.find('.rewind').at(0).simulate('click');
    testRewindGrid.find('.rewind').at(0).simulate('click');
    testRewindGrid.find('.rewind').at(0).simulate('click');
    testRewindGrid.find('.rewind').at(0).simulate('click');
    testRewindGrid.find('.rewind').at(0).simulate('click');
    console.log(testRewindGrid.find('.tic-tac-grid').parent().prop("state")); 

    expect(testRewindGrid.find('.tic-tac-grid').parent().prop("state")).toEqual(initialState);
 });