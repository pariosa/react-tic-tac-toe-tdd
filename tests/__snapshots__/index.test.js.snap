// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `ShallowWrapper {}`;

exports[`renders shallow match 1`] = `ShallowWrapper {}`;
